### PR TITLE
[browser] loader ES6 module

### PR DESCRIPTION
**Pull request by pavelsavara**
Split the `dotnet.js` into 3 ES6 modules.
- new smaller `dotnet.js` 
    - this is "loader", small module which 
    - is produced from code in the `src\mono\wasm\runtime\loader\` folder
    - loads the `mono-config.json` or `boot.config.json`
    - loads `dotnet.runtime.js` and `dotnet.nativejs` by looking at the `config.assets`
- `dotnet.runtime.js` 
    - this is "runtime", the main JS logic we have in typescript
    - is produced from code in the `src\mono\wasm\runtime\` folder except `/loader` folder
    - in the future we could further split off legacy intero and jiterp into separate JS files
- `dotnet.native.js` 
    - is the emscripten module, it changes on rebuild with workload
    - this one also brings the other `dotnet.native.wasm`, `dotnet.native.symbols.js`, `dotnet.native.worker.js`  
    - because `dotnet.native` is the new prefix name.

future benefits
 - we could have hash in names of all of them but the loader, `dotnet.js` . The loader would not change on re-link with workload.
 - "loader" could do feature detection in the future and load different assets instead.
 - further unify/simplify blazor startup code
 - drop DLL cache, because of the hashes

## Other changes
- config has new `assetUniqueQuery` which is appended to asset names (so that we could have multiple instances of ES6 modules)
- `loaderHelpers` is similar to `runtimeHelpers` but is populated by "loader". It passes methods and state between modules.
- `find_corlib_class` and `find_class_in_assembly` would not always throw when they don't find the class.
- updated WBT
- moved files and types
- added rollup assert, so that code doesn't bleed between modules
- added startup assert to throw if the same ES6 module is initialized multiple times
- fixed memory snapshot to play well with the new layout
- new asset types `js-module-runtime`, `js-module-dotnet`, `js-module-native`
- removed `Module.imports` functionality, dropped `DotnetModuleConfigImports`
- removed `MONO.mono_wasm_load_config` from legacy API
- simplified `require` polyfill on NodeJS
- simplified `locateFile` polyfill
- `createEmscripten` is worth review
- fixed `SelfContained`
- worker module `dotnet.native.worker.js` now loads `dotnet.js` loader first

### CODE REVIEW

**Comment by pavelsavara** [comment]
WBT failes with `/root/helix/work/workitem/e/wbt/nuget/lqnz5tnh.1uj/microsoft.net.illink.tasks/8.0.0-preview.5.23260.3/build/Microsoft.NET.ILLink.targets(196,5): error NETSDK1102: Optimizing assemblies for size is not supported for the selected publish configuration. Please ensure that you are publishing a self-contained app.` which I think is unrelated.

All other CI failures are known.

**Review by ilonatommy — APPROVED** [review]
LGTM

File: src/mono/wasm/runtime/dotnet.d.ts [hunk 1]
```diff
@@ -139,6 +139,10 @@ type MonoConfig = {
      * application environment
      */
     applicationEnvironment?: string;
+    /**
+     * query string to be used for asset loading
+     */
+    assetUniqueQuery?: string;
```
Comment (maraf) [inline src/mono/wasm/runtime/dotnet.d.ts:145]: What is the benefit over `WebAssemblyStartOptions.loadBootResource`?

File: src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/ComputeWasmPublishAssets.cs [hunk 2]
```diff
@@ -252,7 +259,7 @@ public override bool Execute()
 
                 ApplyPublishProperties(newDotNetWasm);
                 nativeStaticWebAssets.Add(newDotNetWasm);
-                if (resolvedNativeAssetToPublish.TryGetValue("dotnet.wasm", out var resolved))
+                if (resolvedNativeAssetToPublish.TryGetValue("dotnet.native.wasm", out var resolved))
```
Comment (maraf) [inline src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/ComputeWasmPublishAssets.cs:262]: This needs support for both `dotnet.wasm` and `dotnet.native.wasm`

File: src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/ComputeWasmPublishAssets.cs [hunk 3]
```diff
@@ -268,19 +275,13 @@ public override bool Execute()
...
         }
 
-        static bool IsDotNetWorkerJs(string key)
-        {
-            var fileName = Path.GetFileName(key);
-            return fileName.StartsWith("dotnet.worker.", StringComparison.Ordinal) && fileName.EndsWith(".js", StringComparison.Ordinal);
-        }
-
-        static bool IsDotNetWasm(string key) => string.Equals("dotnet.wasm", Path.GetFileName(key), StringComparison.Ordinal);
+        static bool IsDotNetWasm(string key) => string.Equals("dotnet.native.wasm", Path.GetFileName(key), StringComparison.Ordinal);
```
Comment (maraf) [inline src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/ComputeWasmPublishAssets.cs:284]: This needs support for both `dotnet.wasm` and `dotnet.native.wasm`

File: src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/GenerateWasmBootJson.cs [hunk 4]
```diff
@@ -196,8 +196,7 @@ public void WriteBootJson(Stream output, string entryAssemblyName)
                         string.Equals(assetTraitValue, "native", StringComparison.OrdinalIgnoreCase))
                 {
                     Log.LogMessage(MessageImportance.Low, "Candidate '{0}' is defined as a native application resource.", resource.ItemSpec);
-                    if (string.Equals(fileName, "dotnet", StringComparison.OrdinalIgnoreCase) &&
-                        string.Equals(fileExtension, ".wasm", StringComparison.OrdinalIgnoreCase))
+                    if (fileName.StartsWith("dotnet.native", StringComparison.OrdinalIgnoreCase) && string.Equals(fileExtension, ".wasm", StringComparison.OrdinalIgnoreCase))
```
Comment (maraf) [inline src/tasks/Microsoft.NET.Sdk.WebAssembly.Pack.Tasks/GenerateWasmBootJson.cs:199]: This needs support for both `dotnet.wasm` and `dotnet.native.wasm`

File: src/mono/wasm/runtime/dotnet.d.ts (hunk 1 above)
Comment (pavelsavara) [reply src/mono/wasm/runtime/dotnet.d.ts:145]: We could unify those later. Right now, the `loadBootResource` is not there for non-blazor templates.

**Review by maraf — APPROVED** [review]
Looks good to me 👍

**Comment by lewing** [comment]
This appears to have caused a massive regression in jsimport/export in browserbench?

**Pull request merged**